### PR TITLE
#0: Port all Misc ops to use TensorSpec

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
We need to migrate all ops to use `compute_output_specs` with
TensorSpec, instead of older `compute_output_shapes`

### What's changed
Migrated all misc ops to TensorSpec
Minor infra upgrades to support the migration.

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/12057375995)
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/12133160610)
- [x] [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/12133156506)
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/12146782009)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/pool/upsample/device/upsample_op.cpp [hunk 1]
```diff
@@ -56,55 +56,50 @@ std::vector<tt::tt_metal::LegacyShape> UpSample::compute_output_shapes(const std
...
+        if (input.memory_config().memory_layout == TensorMemoryLayout::BLOCK_SHARDED) {
+            auto shard_grid = input_shard_spec.grid.ranges();
+            TT_FATAL(shard_grid.size() == 1, "Block sharded input should have only one CoreRange");
+            auto core_range = *shard_grid.begin();
+            uint32_t ncores_w = core_range.end_coord.x + 1;
+            uint32_t ncores_h = core_range.end_coord.y + 1;
+            // std::array<uint32_t, 2> output_shard_shape = {output_shape[0] * output_shape[1] * output_shape[2] /
+            // ncores_h, output_shape[-1] / ncores_w}; auto output_shard_spec = input_shard_spec;
+            // output_shard_spec.shape = output_shard_shape;
+            // mem_config.shard_spec = output_shard_spec;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/upsample/device/upsample_op.cpp:89]: remove?

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/pool/downsample/device/downsample_op.cpp [hunk 2]
```diff
@@ -30,27 +30,17 @@ void Downsample::validate(const std::vector<Tensor>& input_tensors) const {
...
+    const auto& input_tensor = input_tensors.at(0);
+    TT_ASSERT(input_tensor.get_legacy_shape()[0] == 1 && input_tensor.get_legacy_shape()[1] == 1);
+    uint32_t input_height = input_tensor.get_legacy_shape()[2];
     auto [img_batch_size, img_height, img_width, img_stride_h, img_stride_w] = this->downsample_params;
     TT_ASSERT(input_height >= img_batch_size * img_height * img_width);
-    uint32_t output_height_unpadded = img_batch_size * ceil((double)img_height / (double)img_stride_h) *
-                                      ceil((double)img_width / (double)img_stride_w);
-    uint32_t output_height = tt::round_up(output_height_unpadded, TILE_HEIGHT);
-    uint32_t output_width = input_tensor_a.get_legacy_shape()[3];
-    auto output_padding =
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/downsample/device/downsample_op.cpp:44]: Seems like this op was trying to encode padding info in it's output tensor, how is handled now?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool/downsample/device/downsample_op.cpp:44]: This computation is not needed anymore. It is now properly handled by tensor layout.

Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/pool/downsample/device/downsample_op.cpp:44]: Ok I'll trust you on that

**Review by pavlejosipovic — APPROVED** [review]

**Pull request merged**